cli_tools/common/image/importer/inflater.go

```diff
@@ -13,40 +13,125 @@
+	Inflate() (persistentDisk, shadowTestFields, error)
+	Cancel(reason string) bool
+}
 
-	// When exceeded, we use default values for PDs, rather than more accurate
-	// values used by inspection. When using default values, the worker may
-	// need to resize the PDs, which requires escalated privileges.
-	inspectionTimeout = time.Second * 3
+type persistentDisk struct {
+	uri        string
+	sizeGb     int64
+	sourceGb   int64
+	sourceType string
+}
 
-	// 10GB is the default disk size used in inflate_file.wf.json.
-	defaultInflationDiskSizeGB = 10
-)
+type shadowTestFields struct {
+	// Below fields are for shadow API inflation metrics
+	checksum      string
+	inflationTime time.Duration
+	inflationType string
+}
+
+func newInflater(request ImageImportRequest, computeClient daisyCompute.Client, storageClient domain.StorageClientInterface,
+	inspector imagefile.Inspector, logger logging.Logger) (Inflater, error) {
+
+	di, err := newDaisyInflater(request, inspector, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	if isImage(request.Source) {
+		return di, nil
+	}
+
+	if isShadowTestFormat(request) {
+		return &shadowTestInflaterFacade{
+			mainInflater:   di,
+			shadowInflater: createAPIInflater(request, computeClient, storageClient, logger, true),
+			logger:         logger,
+		}, nil
+	}
+
+	return &inflaterFacade{
+		apiInflater:   createAPIInflater(request, computeClient, storageClient, logger, false),
+		daisyInflater: di,
+		logger:        logger,
+	}, nil
+}
+
+func isShadowTestFormat(request ImageImportRequest) bool {
+	// TODO: process VHD/VPC differently
+	return false
+}
 
 // inflaterFacade implements an inflater using other concrete implementations.
 type inflaterFacade struct {
+	apiInflater   Inflater
+	daisyInflater Inflater
+	logger        logging.Logger
+}
+
+func (facade *inflaterFacade) Inflate() (persistentDisk, shadowTestFields, error) {
+	var pd persistentDisk
+	var tf shadowTestFields
+	var err error
+
+	//  Try with API inflater.
+	pd, tf, err = facade.apiInflater.Inflate()
+	if err == nil {
+		facade.logger.Metric(&pb.OutputInfo{
+			InflationType:   "api_success",
+			InflationTimeMs: []int64{tf.inflationTime.Milliseconds()},
+		})
+		return pd, tf, err
+	}
+
+	if !isCausedByUnsupportedFormat(err) {
+		facade.logger.Metric(&pb.OutputInfo{
+			InflationType:   "api_failed",
+			InflationTimeMs: []int64{tf.inflationTime.Milliseconds()},
+		})
+		return pd, tf, err
+	}
+
+	// Retry with daisy inflater.
+	pd, tf, err = facade.daisyInflater.Inflate()
+	if err == nil {
+		facade.logger.Metric(&pb.OutputInfo{
+			InflationType:   "qemu_success",
+			InflationTimeMs: []int64{tf.inflationTime.Milliseconds()},
+		})
+		return pd, tf, err
+	}
+
+	facade.logger.Metric(&pb.OutputInfo{
+		InflationType:   "qemu_failed",
+		InflationTimeMs: []int64{tf.inflationTime.Milliseconds()},
+	})
+	return pd, tf, err
+}
+
+func (facade *inflaterFacade) Cancel(reason string) bool {
+	// No need to cancel apiInflater.
```

Comment:
Why do we explicitly have to cancel Daisy and not the API Inflater? Is this something unique to Daisy?

Reply:
The daisy inflater is a workflow and it can be interrupted from the middle. The PD API is one single API call and is pretty fast.  We can cancel it (it's supported), but it's not very valuable to support that - it brings extra complexity while its benefit is small.